classes/core/PKPQueueProvider.php

```diff
@@ -18,53 +18,162 @@
 namespace PKP\core;
 
 use APP\core\Application;
+use Illuminate\Queue\Events\JobFailed;
+use Illuminate\Support\Facades\Queue;
 use Illuminate\Queue\WorkerOptions;
-
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Queue\QueueServiceProvider as IlluminateQueueServiceProvider;
+use Illuminate\Contracts\Debug\ExceptionHandler;
+use Illuminate\Queue\Worker;
 use PKP\config\Config;
 use PKP\Domains\Jobs\Job as PKPJobModel;
+use PKP\Domains\Jobs\Interfaces\JobRepositoryInterface;
+use PKP\Domains\Jobs\Repositories\Job as JobRepository;
+use PKP\Domains\Jobs\WorkerConfiguration;
 
-class PKPQueueProvider
-{
-    public function runJobsAtShutdown(): void
-    {
-        $disableRun = Config::getVar('queues', 'disable_jobs_run_at_shutdown', false);
+class PKPQueueProvider extends IlluminateQueueServiceProvider
+{   
+    /**
+     * Specific queue to target to run the associated jobs
+     */
+    protected ?string $queue = null;
 
-        if ($disableRun || Application::isUnderMaintenance()) {
-            return;
-        }
+    /**
+     * Set a specific queue to target to run the associated jobs
+     */
+    public function forQueue(string $queue) : self
+    {
+        $this->queue = $queue;
+        
+        return $this;
+    }
 
-        $job = PKPJobModel::isAvailable()
-            ->notExceededAttempts()
+    /**
+     * Get a job model builder instance to query the jobs table
+     */
+    public function getJobModelBuilder() : Builder
+    {
+        return PKPJobModel::isAvailable()
             ->nonEmptyQueue()
-            ->notQueue(PKPJobModel::TESTING_QUEUE)
-            ->limit(1)
-            ->first();
+            ->when($this->queue, fn($query) => $query->onQueue($this->queue))
+            ->when(is_null($this->queue), fn($query) => $query->notQueue(PKPJobModel::TESTING_QUEUE))
+            ->notExceededAttempts();
+    }
+
+    /**
+     * Get the worker options object
+     */
+    public function getWorkerOptions(array $options = []): WorkerOptions
+    {
+        return new WorkerOptions(...WorkerConfiguration::withOptions($options)->getWorkerOptions());
+    }
+
+    /**
+     * Run the queue worker via an infinite loop daemon
+     */
+    public function runJobsViaDaemon(string $connection, string $queue, array $workerOptions = []): void
+    {
+        $laravelContainer = PKPContainer::getInstance();
+
+        $laravelContainer['queue.worker']->daemon(
+            $connection,
+            $queue,
+            $this->getWorkerOptions($workerOptions)
+        );
+    }
+    
+    /**
+     * Run the queue worker to process queue the jobs
+     */
+    public function runJobsInQueue(): void
```

Comment:
This function only runs a single job, right? `runJobsInQueue` might be misleading.

Reply:
right. as it's handle one job at a time , it should be in singular like `runJobInQueue`. Done fixing .